### DISCUSSION TITLE
How to create a websocket endpoint?

### DISCUSSION THREAD

**Discussion by Irvingouj** [pocketbase/pocketbase#4917] [Q&A]
How to create a websocket endpoint on custom router? I am trying to implement a video call, where the server and client require realtime data transfer, while the realtime api is great but it limits my ability to provide customizations.

Your response is apperciated!

**Comment by lukepighetti** [comment]
What limitations are you experiencing with SSE real-time events?

**Reply by erickythierry** [reply]
I was thinking about this idea of ​​an endpoint WebSocket as well, and I believe that the limitation of the current PocketBase Realtime method is the lack of a way to detect customer connections/disconnections at the time, without the need for client to inform this (whether modifying a registration in DB or sending a request).

I did not get deep as it works for this Real-Time Events, but I imagine it does not have this capacity that the WebSocket has to detect this type of event on the side of the server the moment they occur.

I am developing a multiple instance system using Docker Swarm to replicate Backend, and with that I needed a "orchestrator" that will identify the instances that are working and distribute the sessions for them, and if one of the backbound replicas stops Working (crash), it would be possible to detect through the Websocket, since the server would receive the event that this instance was disconnected, so would make the logic of distributing the sessions of this disconnected server to the others (or move to new instance that climb in place of the previous one).

Do you know some way to do this using the SSE Real-Time Events?

I even imagined creating an extra server to serve as an orchestrator, but it would be something more to take care of and keep running, so I thought of leaving it straight to DB (pocketbase)

**Reply by ganigeorgiev** [reply]
@erickythierry I wouldn't have responded to this comment but I'd appreciate if users don't speak so confident about things they don't understand because often it is half-true or plain wrong and others tend to repeat it (_the same reason why if you Google "PocketBase transactions" one of the first search results would be some silly reddit post claiming that PocketBase doesn't support them, which is obviously wrong and `RunInTransaction` exists since the first release; similar situation with cursor-based pagination, but anyway_).

> I believe that the limitation of the current PocketBase Realtime method is the lack of a way to detect customer connections/disconnections at the time, without the need for client to inform this (whether modifying a registration in DB or sending a request).

You are mistaken. If you use the [Go or JS realtime hooks](https://pocketbase.io/docs/go-event-hooks/#realtime-api-hooks) you have information when clients connect/disconnect and you can notify other users (`app.SubscriptionsBroker().Clients()`) about it by sending them a custom message.

The problem with "presence"-like functionality is not the SSE implementation, but the lack of clarity who should have access to that information. It is difficult to generalize and because of this it is left to users to handle on their own based on their specific use case (the topic was discussed couple times already in https://github.com/pocketbase/pocketbase/discussions/1706, https://github.com/pocketbase/pocketbase/discussions/1607, https://github.com/pocketbase/pocketbase/discussions/3378 and probably others).

But in any case this is off-topic to the original question. If the linked discussions doesn't help, consider opening a new one with more details about your use case and what you are trying to do.

**Reply by erickythierry** [reply]
> @erickythierry I wouldn't have responded to this comment but I'd appreciate if users don't speak so confident about things they don't understand because often it is half-true or plain wrong and others tend to repeat it (_the same reason why if you Google "PocketBase transactions" one of the first search results would be some silly reddit post claiming that PocketBase doesn't support them, which is obviously wrong and `RunInTransaction` exists since the first release; similar situation with cursor-based pagination, but anyway_).
> 
> > I believe that the limitation of the current PocketBase Realtime method is the lack of a way to detect customer connections/disconnections at the time, without the need for client to inform this (whether modifying a registration in DB or sending a request).
> 
> You are mistaken. If you use the [Go or JS realtime hooks](https://pocketbase.io/docs/go-event-hooks/#realtime-api-hooks) you have information when clients connect/disconnect and you can notify other users (`app.SubscriptionsBroker().Clients()`) about it by sending them a custom message.
> 
> The problem with "presence"-like functionality is not the SSE implementation, but the lack of clarity who should have access to that information. It is difficult to generalize and because of this it is left to users to handle on their own based on their specific use case (the topic was discussed couple times already in #1706, #1607, #3378 and probably others).
> 
> But in any case this is off-topic to the original question. If the linked discussions doesn't help, consider opening a new one with more details about your use case and what you are trying to do.

I see. As I said, I didn't go into depth about how SSE works in PocketBase. I limited myself to what is officially available in the REST endpoints and in the JS SDK. And since I didn't find anything related to this type of event, I thought it wasn't possible to do it, but I was wrong, I admit.
I made my comment in this post precisely so I wouldn't have to open a new one with the same focus and doubt, since I also had this idea of ​​whether it was possible to create a websocket endpoint.

Regarding this PocketBase capability that you mentioned, app.SubscriptionsBroker().Clients(), would there be documentation about it? Because I'm really interested in adding this logic to my project and keeping it in PocketBase.
